src/charm.py

```diff
@@ -0,0 +1,145 @@
+        """
+        super().__init__(framework)
+        framework.observe(self.on.start, self._on_start)
+        framework.observe(self.on.stop, self._on_stop)
+        framework.observe(self.on.update_status, self._on_update_status)
+        framework.observe(
+            self.on[CACHE_CONFIG_INTEGRATION_NAME].relation_changed,
+            self._on_cache_config_relation_changed,
+        )
+        framework.observe(
+            self.on[CACHE_CONFIG_INTEGRATION_NAME].relation_broken,
+            self._on_cache_config_relation_broken,
+        )
+
+    def _on_start(self, _: ops.StartEvent) -> None:
+        """Handle start event."""
+        _nginx_initialize()
+        self._set_status()
+
+    def _on_stop(self, _: ops.StopEvent) -> None:
+        """Handle the stop event."""
+        _nginx_stop()
+
+    def _on_update_status(self, _: ops.UpdateStatusEvent) -> None:
+        """Handle update status event."""
+        self._set_status()
+
+    def _on_cache_config_relation_changed(self, _: ops.RelationChangedEvent) -> None:
+        """Handle config relation changed event."""
+        self._load_nginx_config()
+
+    def _on_cache_config_relation_broken(self, _: ops.RelationBrokenEvent) -> None:
+        """Handle config relation broken event."""
+        self._load_nginx_config()
+
+    def _set_status(self) -> None:
+        """Set the charm status."""
+        if not self.model.relations[CACHE_CONFIG_INTEGRATION_NAME]:
+            self.unit.status = ops.BlockedStatus(WAIT_FOR_CONFIG_MESSAGE)
+            return
+
+        if not nginx_manager.ready_check():
+            self.unit.status = ops.MaintenanceStatus(NGINX_NOT_READY_MESSAGE)
+            return
+
+        self.unit.status = ops.ActiveStatus()
+
+    def _load_nginx_config(self) -> None:
+        """Validate the configuration and load to integration.
+
+        Raises:
+            NginxFileError: File operation errors while updating nginx configuration files.
+        """
+        try:
+            nginx_config = get_nginx_config(self)
+        except IntegrationDataError as err:
+            self.unit.status = ops.BlockedStatus(str(err))
+            return
+        if not nginx_config:
+            _nginx_stop()
+            self.unit.status = ops.BlockedStatus(WAIT_FOR_CONFIG_MESSAGE)
+            return
+
+        status_message = ""
+        try:
+            nginx_manager.update_config(nginx_config)
+        except NginxFileError:
+            logger.exception(
+                "Failed to update nginx config file, going to error state for retries"
+            )
+            raise
+        except NginxConfigurationAggregateError as err:
+            logger.exception("Found error with configuration for hosts: %s", err.hosts)
+            logger.warning(
+                "Any hosts configuration without errors will be served on content cache"
+            )
+            status_message = f"Error for host: {err.hosts}"
+
+        nginx_manager.load_config()
```

Comment:
I was thinking about the interface, couldn't the update and load config be the same function? What would be the advantages of separating them?

Reply:
There is a bit of discussion here: https://github.com/canonical/content-cache-operator/pull/2#discussion_r1796552668

Reply:
Hmm, what would be the issue of having a function `update_and_load_config(config)`? Would we ever update the config without the immediately loading it?

Reply:
I do not think of any case the would involve update and load config separately.

However, I feel splitting off the `load_config` in its own function might be fine, since the update config part will get more complex. For the next PR I am working on, I have split of update configuration to different single-usage functions due to the linting complaining this function is too complex.